import sys
sys.path.insert(0, '/srv/biodata/dep_mercier/grp_schneeberger/software/hometools')
from myUsefulFunctions import readfasta, readblast

def filter_orthologs(geneids, orthogroups):
    """
    Get Currot genes that are orthologous to the cell-type specific marker genes from AThal.
    geneids: file contatining list of genes whose orthologs are to be selected
    orthogroups: path to Orthogroups.txt file generated by OrthoFinder
    """
    from collections import deque
    import re
    genes = [l.strip().split()[0] for l in open(geneids, 'r')]
    ortho = deque()
    for l in open(orthogroups, 'r'):
        for gene in genes:
            if len(re.findall(gene+'.', l)) > 0:
                l2 = l.strip().split()
                for g in l2[1:]:
                    if g[:2] != 'AT':
                        ortho.append(g)
                break
    ortho = list(set(ortho))
    print(*ortho, sep='\n')


def filter_cds_blast(blastout, queryseq, marority=0.75):
    """
    Filter out CDS blast output to select only matching covering majority of the
    query
    """
    qryseq = {k: len(v) for k, v in readfasta(queryseq).items()}

